sdk/storage/azure-storage-blob-nio/src/test/java/com/azure/storage/blob/nio/CompositeTest.groovy

```diff
@@ -106,4 +112,130 @@ class CompositeTest extends APISpec {
+                putDirectoryBlob(blobClient.getBlockBlobClient())
+            }
+            childClient1.upload(data.defaultInputStream, data.defaultDataSize)
+        }
+
+        expect:
+        if (status != DirectoryStatus.DOES_NOT_EXIST) {
+            assert Files.exists(path)
+        } else {
+            assert !Files.exists(path)
+        }
+
+        where:
+        status                          | isVirtual
+        DirectoryStatus.DOES_NOT_EXIST  | false
+        DirectoryStatus.NOT_A_DIRECTORY | false
+        DirectoryStatus.NOT_EMPTY       | true
+        DirectoryStatus.NOT_EMPTY       | false
+    }
+
+    def "Files walkFileTree"() {
+        setup:
+        def fs = createFS(config)
+        /*
+        file1
+        cDir1
+        cDir2
+        |__file2
+        |__cDir3
+        |__vDir1
+           |__file3
+        vDir2
+        |__file4
+        |__cDir4
+        |__vDir3
+           |__file5
+         */
+        def baseDir = "a"
+        def file1 = "a/file1"
+        def cDir1 = "a/cDir1"
+        def cDir2 = "a/cDir2"
+        def file2 = "a/cDir2/file2"
+        def cDir3 = "a/cDir2/cDir3"
+        def vDir1 = "a/cDir2/vDir1"
+        def file3 = "a/cDir2/vDir1/file3"
+        def vDir2 = "a/vDir2"
+        def file4 = "a/vDir2/file4"
+        def cDir4 = "a/vDir2/cDir4"
+        def vdir3 = "a/vDir2/vDir3"
+        def file5 = "a/vDir2/vDir3/file5"
+
+        // Create files and directories
+        ((AzurePath) fs.getPath(file1)).toBlobClient().upload(data.defaultInputStream, data.defaultDataSize)
+        ((AzurePath) fs.getPath(file2)).toBlobClient().upload(data.defaultInputStream, data.defaultDataSize)
+        ((AzurePath) fs.getPath(file3)).toBlobClient().upload(data.defaultInputStream, data.defaultDataSize)
+        ((AzurePath) fs.getPath(file4)).toBlobClient().upload(data.defaultInputStream, data.defaultDataSize)
+        ((AzurePath) fs.getPath(file5)).toBlobClient().upload(data.defaultInputStream, data.defaultDataSize)
+
+        putDirectoryBlob(((AzurePath) fs.getPath(baseDir)).toBlobClient().getBlockBlobClient())
+        putDirectoryBlob(((AzurePath) fs.getPath(cDir1)).toBlobClient().getBlockBlobClient())
+        putDirectoryBlob(((AzurePath) fs.getPath(cDir2)).toBlobClient().getBlockBlobClient())
+        putDirectoryBlob(((AzurePath) fs.getPath(cDir3)).toBlobClient().getBlockBlobClient())
+        putDirectoryBlob(((AzurePath) fs.getPath(cDir4)).toBlobClient().getBlockBlobClient())
+
+        when:
+        def visitor = new TestFileVisitor()
+        System.out.println(Files.readAttributes(fs.getPath(baseDir), AzureBasicFileAttributes).isDirectory())
+        Files.walkFileTree(fs.getPath(baseDir), visitor)
+
+        // might need to make this work on root directories as well, which would probably mean inspecting the path and adding an isRoot method
+
+        then:
+        visitor.fileCount == 5
+        visitor.directoryCount == 8 // includes baseDir
+        visitor.failureCount == 0
```

Comment:
should we validate that visitor is observing expected paths ?

Reply:
This felt like overkill to me because it seemed like it'd be more validating that the jvm's implementation of walkFileTree is correct than that we have the correct behavior to enable proper use of that method. If you don't agree with that, though, I can add in validation that we hit all the expected paths.